fish-rust/src/wutil/mod.rs

```diff
@@ -68,13 +68,13 @@ pub fn wperror(s: &wstr) {
 
 /// Port of the wide-string wperror from `src/wutil.cpp` but for rust `&str`.
 pub fn perror(s: &str) {
-    let e = errno().0;
+    let e = Errno::last();
```

Comment:
The weird thing about `nix::errno::Errno` is that they treat "unknown error" the same as "no error".
This could cause subtle issues if the OS ever returns an error that hasn't been added to `nix` yet.
Unless they have a way of guaranteeing consistency with the host libc?

---

I'm not sure `enum` is the right type for `Errno`.
It's already marked `#[non_exhaustive]` so client code can't do exhaustive matching.
I guess they could make it:

```
enum Errno {
    NoError = 0
    EPERM = libc::EPERM,
    ...
    UnknownErrno(i32),
}
```

but at that point it's probably better to use a struct that wraps `libc::c_int` (or `i32`?), similar to `std::io::Error`.

---

I guess we could use the `nix::errno::Errno::last_raw()` (which is `nix::errno::errno()` today) and `nix::errno::Errno::set_raw()` precisely in the cases where we don't want to ignore unknown errors.
But the problem of being easy to misuse remains.

Reply:
Yeha, I noticed that as well, it may cause problems in places where we do if errno != 0. Where unknown error will be treated as success.

The best way would be to remove != 0 checks in favour of Result.is_ok, and in places where we actually need to fiddle with errnos do the raw ones.

So simply, we would not use NoErrno anywhere and assume that we don't care about the difference between unsupported errno types, would that be fine?

Obviously I would be open to working with nix to perhaps add a way to distinguish those, they probably simply never needed that as they do Results instead of NoErrno values everywhere.

Reply:
> Yeha, I noticed that as well, it may cause problems in places where we do if errno != 0

Right, this is actually a rare case so it does little harm.

> So simply, we would not use NoErrno anywhere and assume that we don't care about the difference between unsupported errno types, would that be fine?

Sounds good. To be clear, cases like `perror` that handle arbitrary errors should use the raw errno, just in case.

It's probably wise to first convert things like `wrealpath` to return `Result`, to make it easy to review.

Reply:
> It's probably wise to first convert things like wrealpath to return Result, to make it easy to review.

Yeah, this is a better approach tbh, I think I will draft this PR for now, and in separate smaller PRs focus on moving stuff to nix & Results rather than focusing on removing errno crate.

After that is done, I will come back to the topic of errno removal.

Reply:
whatever works. We can start merging all the uses of errno where we only compare against concrete errors

Reply:
Hi, one of the Nix maintainers here, happy to help to make Nix serve fish better!

> The weird thing about nix::errno::Errno is that they treat "unknown error" the same as "no error".
This could cause subtle issues if the OS ever returns an error that hasn't been added to nix yet.

Yeah, this is indeed a bad design, I think we can do the following approach:
>
> ```rs
> enum Errno {
>     NoError = 0
>     EPERM = libc::EPERM,
>      ...
>     UnknownErrno(i32),
> }
> ```

But maybe `NoError` should not be included as one of them, and for `nix::errno::Errno::last()`, we can make it return an `Option<Errno>`:

```rs
impl Errno {
    pub fn last() -> Option<Errno> {
        match Self::last_raw() {
            0 => None,
            raw => Some(Self::from_raw(raw)),
        }
    }
}
```